python/pyarrow/src/arrow/python/udf.cc

```diff
@@ -234,6 +351,56 @@ Status RegisterTabularFunction(PyObject* user_function, UdfWrapperCallback wrapp
       wrapper, options, registry);
 }
 
+Status AddAggKernel(std::shared_ptr<compute::KernelSignature> sig, compute::KernelInit init,
+                           compute::ScalarAggregateFunction* func) {
+
+  compute::ScalarAggregateKernel kernel(std::move(sig), std::move(init), AggregateUdfConsume, AggregateUdfMerge, AggregateUdfFinalize, /*ordered=*/false);
+  RETURN_NOT_OK(func->AddKernel(std::move(kernel)));
+  return Status::OK();
+}
+
+Status RegisterAggregateFunction(PyObject* agg_function, UdfWrapperCallback agg_wrapper,
+						const UdfOptions& options,
+						compute::FunctionRegistry* registry) {
+  if (!PyCallable_Check(agg_function)) {
+    return Status::TypeError("Expected a callable Python object.");
+  }
+
+  if (registry == NULLPTR) {
+    registry = compute::GetFunctionRegistry();
+  }
+
+  static auto default_scalar_aggregate_options = compute::ScalarAggregateOptions::Defaults();
+  auto aggregate_func = std::make_shared<compute::ScalarAggregateFunction>(
+      options.func_name, options.arity, options.func_doc, &default_scalar_aggregate_options);
+
+  Py_INCREF(agg_function);
```

Comment:
Why is this needed?  Can we wrap `agg_function` in an `OwnedRef` instead?

Reply:
I am actually no sure about how Py reference counting should work so I think I could use some help here. Currently I wrote this by trial and error and found if I don't have the Py_INCREF here it would segfault. 

@westonpace What is the best way for me to understand how Py_INCREF and OwnedRef are supposed to work here?

Reply:
I am actually no sure about how Py reference counting should work so I think I could use some help here. Currently I wrote this by trial and error and found if I don't have the Py_INCREF here it would segfault. 

@westonpace What is the best way for me to understand how Py_INCREF and OwnedRef are supposed to work here?

Reply:
My current probably incorrect understanding is that if we are passing a PyOjbect to a OwnedRef or OwnedRefNoGIL then we need to call Py_INCREF because the de constructor of OwnedRef and OwnedRefNoGIL will Py_XDECREF?

Reply:
I looked into this more and observed the following.

The refcount of Python function is 2 before registering with Arrow, and is 3 after registering. After invoking the function with arrow compute, the refcount doesn't increase. (still 3) So I believe the code is correct.

On the other hand, I did observe scalar UDF might have a refcount leaking (I observed the refcount increases by 1 after calling the function with arrow compute). Which I can look into separately.

In summary, my understanding is that every time we wrap the PyObject* into a OwnedRef, we need to increase the refcount by 1 so the refcount doesn't change after the destructor of OwnedRef gets invoked. I am 90% certain this is correct understanding of how it should work but happy to hear more thoughts @westonpace